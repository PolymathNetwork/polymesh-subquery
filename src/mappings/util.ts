import { Codec } from "@polkadot/types/types";
import { u8aToString } from "@polkadot/util";
/**
 * @returns a javascript object built using an `iterable` of keys and values.
 * Values are mapped by the map parameter
 */
export const fromEntries = <K extends string | number, V, V2>(
  iterable: Iterable<[K, V]>,
  map: (v: V, i: number, k: K) => V2
): Partial<Record<K, V2>> => {
  const res: Partial<Record<K, V2>> = {};

  let i = 0;
  for (const [k, v] of iterable) {
    res[k] = map(v, i, k);
    i++;
  }
  return res;
};

export const camelToSnakeCase = (str: string): string =>
  str.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);

export const capitalizeFirstLetter = (str: string): string =>
  str[0].toUpperCase() + str.slice(1);

export const removeNullChars = (s: string): string => s.replace(/\0/g, "");

/**
 * @returns the index of the first top level comma in `text` which is a string with nested () and <>.
 * This is meant for rust types, for example "Map<Map<(u8,u32),String>,bool>"" would return 24.
 * @param text
 */
export const findTopLevelCommas = (
  text: string,
  exitOnFirst = false
): number[] => {
  let nestedLevel = 0;
  const commas = [];
  let i = 0;

  for (const char of text) {
    switch (char) {
      case "(":
      case "{":
      case "<": {
        nestedLevel++;
        break;
      }
      case ")":
      case "}":
      case ">": {
        nestedLevel--;
        break;
      }
      case ",": {
        if (nestedLevel === 1) {
          if (exitOnFirst) {
            return [i];
          } else {
            commas.push(i);
          }
        }
      }
    }
    i++;
  }
  if (commas.length === 0) {
    throw new Error(
      `No top level comma found in ${text}, it probably isn't a map`
    );
  }
  return commas;
};

export const getOrDefault = <K, V>(
  map: Map<K, V>,
  key: K,
  getDefault: () => V
): V => {
  const v = map.get(key);
  if (v !== undefined) {
    return v;
  } else {
    const def = getDefault();
    map.set(key, def);
    return def;
  }
};

export const serializeTicker = (item: Codec): string => {
  return removeNullChars(u8aToString(item.toU8a()));
};
