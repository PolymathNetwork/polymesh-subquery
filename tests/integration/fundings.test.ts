import { gql } from '@apollo/client/core';
import { getApolloClient } from '../util';
const { query } = getApolloClient();

describe('getFundings', () => {
  test('test getting fundings with ticker filter only 1', async () => {
    const q = {
      query: gql`
        query {
          fundings(filter: { ticker: { equalTo: "8TICKER" } }, first: 10) {
            totalCount
            nodes {
              blockId
              ticker
              fundingName
              value
              totalIssuedInFundingRound
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });

  test('test getting fundings with ticker and fundingName filters 1', async () => {
    const q = {
      query: gql`
        query {
          fundings(
            filter: { fundingName: { equalTo: "first" }, ticker: { equalTo: "8TICKER" } }
            first: 10
          ) {
            totalCount
            nodes {
              blockId
              ticker
              fundingName
              value
              totalIssuedInFundingRound
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });

  test('test getting fundings with ticker and fundingName filters 2', async () => {
    const q = {
      query: gql`
        query {
          fundings(
            filter: { fundingName: { equalTo: "" }, ticker: { equalTo: "8TICKER" } }
            first: 10
          ) {
            totalCount
            nodes {
              blockId
              ticker
              fundingName
              value
              totalIssuedInFundingRound
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });
});
