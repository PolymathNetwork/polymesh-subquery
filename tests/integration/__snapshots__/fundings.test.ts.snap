// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFundings test getting fundings with ticker and fundingName filters 1 1`] = `
Object {
  "fundings": Object {
    "__typename": "FundingsConnection",
    "nodes": Array [
      Object {
        "__typename": "Funding",
        "blockId": "201",
        "fundingName": "first",
        "ticker": "8TICKER",
        "totalIssuedInFundingRound": "1000000",
        "value": "1000000",
      },
      Object {
        "__typename": "Funding",
        "blockId": "204",
        "fundingName": "first",
        "ticker": "8TICKER",
        "totalIssuedInFundingRound": "1000100",
        "value": "100",
      },
    ],
    "totalCount": 2,
  },
}
`;

exports[`getFundings test getting fundings with ticker and fundingName filters 2 1`] = `
Object {
  "fundings": Object {
    "__typename": "FundingsConnection",
    "nodes": Array [],
    "totalCount": 0,
  },
}
`;

exports[`getFundings test getting fundings with ticker filter only 1 1`] = `
Object {
  "fundings": Object {
    "__typename": "FundingsConnection",
    "nodes": Array [
      Object {
        "__typename": "Funding",
        "blockId": "201",
        "fundingName": "first",
        "ticker": "8TICKER",
        "totalIssuedInFundingRound": "1000000",
        "value": "1000000",
      },
      Object {
        "__typename": "Funding",
        "blockId": "204",
        "fundingName": "first",
        "ticker": "8TICKER",
        "totalIssuedInFundingRound": "1000100",
        "value": "100",
      },
    ],
    "totalCount": 2,
  },
}
`;
